xilem_web/src/memoized_await.rs

```diff
@@ -0,0 +1,248 @@
+// Copyright 2024 the Xilem Authors and the Druid Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{future::Future, marker::PhantomData};
+
+use wasm_bindgen::{closure::Closure, JsCast, UnwrapThrowExt};
+use wasm_bindgen_futures::spawn_local;
+use xilem_core::{MessageResult, Mut, NoElement, View, ViewId, ViewPathTracker};
+
+use crate::{DynMessage, OptionalAction, ViewCtx};
+
+/// Await a future returned by `init_future` invoked with the argument `data`, `callback` is called with the output of the future. `init_future` will be invoked again, when `data` changes. Use [`memoized_await`] for construction of this [`View`]
+pub struct MemoizedAwait<State, Action, OA, InitFuture, Data, Callback, F, FOut> {
+    init_future: InitFuture,
+    data: Data,
+    callback: Callback,
+    debounce: usize,
+    reset_debounce_on_update: bool,
+    #[allow(clippy::type_complexity)]
+    phantom: PhantomData<fn() -> (State, Action, OA, F, FOut)>,
+}
+
+impl<State, Action, OA, InitFuture, Data, CB, F, FOut>
+    MemoizedAwait<State, Action, OA, InitFuture, Data, CB, F, FOut>
+{
+    /// Debounce the `init_future` function, when `data` updates,
+    /// when `reset_debounce_on_update == false` then this throttles updates each `millisecconds`
+    pub fn debounce(mut self, milliseconds: usize) -> Self {
```

Comment:
So this doesn't debounce when this view is spawned (e.g. as in example `(state.cats_to_fetch < TOO_MANY_CATS).then_some(memoized_await(..))`), I'm not sure whether it actually *should* debounce in `View::build`, so that it has the same behavior when it's in the view tree, as well as when it's being spawned.

Reply:
I've did the debounce now in `View::build` and slightly refactored the `View` impl as well, as I think this should be able to be controlled.

When this behavior is not desired, it could be disabled by the user with `.debounce(0)` at the beginning and reenabled in the future callback (not that this is super awesome, but it at a least allows the user to decide). In the example it "feels" better with having the debounce in the beginning (when going over the `TOO_MANY_CATS` boundary).